### DISCUSSION TITLE
Properties questions

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#10097] [Q&A]
I have read about direct and styled properties in the main docs.  I have a few questions.

1. In choosing between styled and direct properties, it appears that if I want to do extra stuff in the setter, then I should use direct properties, right?  The tradeoff is the prioritized styling that a styled property offers, and some other overhead is saved?

2.  Both styled and direct properties are accessible as attributes in axaml.  Given the documented differences, is choosing one or the other a matter of whether I care that styles are prioritized, I'm guessing when you have multiple layers of styles trying to manipulate the value of the property?

2.  How can I make it so a (Avalonia?) property is *not* accessible from the xaml, but still emits observable events for the UI to pick up?  Do I just use a regular C# property?  Or some other observable type of object?  Not sure the pattern here.

BTW Avalonia is great.  Thank you all devs for the hard work and fun project.

**Comment by maxkatz6** [comment]
> it appears that if I want to do extra stuff in the setter, then I should use direct properties, right?

Technically, you should do this extra stuff in the OnPropertyChanged overload/event.
Because it's still possible to set styled property value with `control.SetValue(Property, value)` which will ignore setter completely. It's not the case with direct properties though.

> is choosing one or the other a matter of whether I care that styles are prioritized,

Styled properties can hold multiple values at once with different priorities, yes.
Direct property just wraps a normal C# property with a single value possible.
That's a main difference.

There are some practical differences as well. I.e. you can only use StyledProperty in the styles (as these need to add and remove value from the object, without affecting previous values). Or you can only use DirectProperty for the readonly properties or with validation (I suppose validation will be added to the StyledProperties at some point too).

> How can I make it so a (Avalonia?) property is not accessible from the xaml, but still emits observable events for the UI to pick up? Do I just use a regular C# property? Or some other observable type of object? Not sure the pattern here.

I am not sure if you can.
From XAML you can even set value to a regular C# property. 
There is no way to hide property. Why?

**Reply by maxkatz6** [reply]
You can hide property by making it private/internal, but it also will hide it from the C# too.

**Reply by sonicsmooth** [reply]
Thanks.  The specific use case here is that I'm keeping track of the pointer position by updating a property with the value I get from OnPointerMoved.  This is captured, then multiplied with the inverse of the transform matrix, and the value displayed in the UI , eg `<Label Content="{Binding #GraphGrid.PointerPosUnit}"/>`  There are also other internal things which are calculated that I wish to display on the UI, but that shouldn't really be settable by any xaml code.

When I instantiate the control, the Intellisense shows this property, indicating that I *could* write this in the xaml as an attribute, thus setting the value of where the mouse is.  But this doesn't make sense, so I don't think this type of thing should be exposed as an attribute, only as a binding.

![image](https://user-images.githubusercontent.com/672429/214963007-bd9e06e9-b5f1-4270-a985-db8e669b1f22.png)

So I'd like to be able to compute stuff in my codebehind and have it be bound to in the UI, without being able to set its value forcibly via xaml attribute.

I tried making this a normal, field-backed property, using `this.RaisAndSetIfChanged` in the setter, but evidently I have to be in a `ViewModelBase` or derived from `ReactiveObject`.  It seems like I'm missing something basic here.

thank you

**Reply by timunie** [reply]
Sounds like a Readonly property to me. It's basically a direct property omitting the setter.

**Reply by maxkatz6** [reply]
If this property has a private setter, but still visible in the AvaloniaVS plugin, then you should create an issue in Avalonia/AvaloniaVS repository.
If there is a public setter, then I don't think we can do anything. As this property is settable from C# and XAML.

**Reply by sonicsmooth** [reply]
I changed the setter to private.  So now intellisense doesn't see it in xaml.  So that's good!  But I can still build and there is no complaint from compiler. (ok but it crashes at runtime -- good!).  I need to learn more about c# observables.

GraphGrid.axaml.cs codebehind:

    public static readonly DirectProperty<GraphGrid, Point> PointerPosPxProperty =
            AvaloniaProperty.RegisterDirect<GraphGrid, Point>(
                nameof(PointerPosPx),
                o => o.PointerPosPx);
    private Point _pointerPosPx = new Point(0, 0);
    public Point PointerPosPx
    {
        get => _pointerPosPx;
        private set => this.SetAndRaise(PointerPosPxProperty, ref _pointerPosPx, value);
    }

MainWindow.axaml

    <controls:GraphGrid Grid.Row="1"
                        Grid.Column="1"
                        Name="GraphGrid"
                        Classes="MainGraphGrid"
                        PointerPosPx="0,0"/>

**Reply by maxkatz6** [reply]
I think in 11.0 it will properly stop the build if there is private property used. Not sure if I remember that correctly though.
